src/step.rs

```diff
@@ -95,10 +98,10 @@ impl State {
     }
 }
 
-impl Iterator for State {
-    type Item = State;
+impl Iterator for Step {
+    type Item = Step;
 
-    fn next(&mut self) -> Option<State> {
+    fn next(&mut self) -> Option<Step> {
```

Comment:
Very strange but it seems that implementing `Iterator` over `Enum` is a footgun as it messes with `partial_cmp` as it expects `IntoIterator`. For example, this test will fail:

```rust
#[cfg(test)]
mod tests {
    pub use super::*;

    #[test]
    pub fn test1() {
        assert_eq!(Step::Prevote.partial_cmp(Step::NewRound),Some(Ordering::Greater));
        assert_eq!(Step::Prevote.partial_cmp(Step::Prevote),Some(Ordering::Equal));
        assert_eq!(Step::Prevote.partial_cmp(Step::Commit),Some(Ordering::Less));
    }
}
```

Reply:
That is odd. The partialord is manually implemented, so this test case shouldn't fail. Checking that rn

Reply:
The problem is with the function resolution. As you pointed, this will fail

```rust
    assert_eq!(
        Step::Prevote.partial_cmp(Step::Commit),
        Some(Ordering::Less)
    );
```

But this will pass

```rust
    assert_eq!(
        std::cmp::PartialOrd::partial_cmp(&Step::Prevote, &Step::Commit),
        Some(Ordering::Less)
    );
```

This happens because the compiler resolves first the [Iterator::partial_cmp](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partial_cmp), so we need to specify the resolver trait there. If you ask me, it was a poor design choice of the iterator trait in the stdlib since now we have this conflict.